rq/worker.py

```diff
@@ -1629,3 +1643,208 @@ class RandomWorker(Worker):
+                            continue
+                        self.log.info("Worker %s: finished executing %d jobs, quitting", self.key, completed_jobs)
+                        break
+
+                except redis.exceptions.TimeoutError:
+                    self.log.error(f"Worker {self.key}: Redis connection timeout, quitting...")
+                    break
+
+                except StopRequested:
+                    break
+
+                except SystemExit:
+                    # Cold shutdown detected
+                    raise
+
+                except:  # noqa
+                    self.log.error('Worker %s: found an unhandled exception, quitting...', self.key, exc_info=True)
+                    break
+        finally:
+            self.register_death()
+            self.unsubscribe()
+
+        return bool(completed_jobs)
+
+    def execute_job(self, job, queue):
+        def job_done(future: Future):
+            """Callback function that runs after the job (future) is finished.
+            This will update the `idle_counter` object and update the `_current_jobs` array,
+            removing the job that just finished from the list.
+
+            Args:
+                future (Future): The Future object.
+            """
+            self.__change_idle_counter(+1)
+            self.heartbeat()
+            job_element = list(filter(lambda x: id(x[1]) == id(future), self._current_jobs))
+            for el in job_element:
+                self._current_jobs.remove(el)
+            if job.get_status() == JobStatus.FINISHED:
+                queue.enqueue_dependents(job)
+
+        self.log.info("Executing job %s from %s", blue(job.id), green(queue.name))
+        future = self.executor.submit(self.perform_job, job, queue)
+        self._current_jobs.append((job, future))
+        self.__change_idle_counter(-1)
+        future.add_done_callback(job_done)
+
+    def wait_all(self, timeout: Optional[int] = None):
+        """ Wait all current jobs """
+        wait([future for _, future in self._current_jobs])
+
+    def __change_idle_counter(self, operation: int):
+        """Updates the idle threads counter using a lock to make it safe.
+
+        Args:
+            operation (int): +1 to increment, -1 to decrement.
+        """
+        with self._lock:
+            self._idle_threads += operation
+
+    def __wait_for_slot(self, wait_interval: float = 0.25):
+        """Waits for a free slot in the thread pool.
+        Sleeps for `wait_interval` seconds to avoid high CPU burden on long jobs.
+
+        Args:
+            wait_interval (float, 0.25): How long to wait between each check. Default to 0.25 second.
+        """
+        while 1:
+            if not self.is_pool_full:
+                self.log.debug('Found idle thread, ready to work')
+                break
+            time.sleep(wait_interval)
+            continue
+
+    def __check_pending_dependents(self) -> bool:
```

Comment:
I think we need to restructure `work()` in such a way that this is not needed.

Reply:
The reason I went this path here is that in this case, things don't happen sequentially, so we need a global way of handling dependencies (thread Y can't work unless thread X has finished) - this is not a challenge on the `Worker` and on the `SimpleWorker`.